### PR TITLE
IGNITE-13126: Reduce continuous query heap usage

### PR DESCRIPTION
**Pull request by Mmuzaf**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java [hunk 1]
```diff
@@ -76,80 +78,58 @@
...
-
-        while (it.hasNext()) {
-            CacheContinuousQueryEntry backupEntry = it.next();
-
-            if (backupEntry.updateCounter() <= updateCntr)
-                it.remove();
-        }
-
+    void cleanupOnAck(long updateCntr) {
+        backupQ.removeIf(backupEntry -> backupEntry.updateCounter() <= updateCntr);
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java:100]: Add line break, please

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryPartitionRecovery.java [hunk 2]
```diff
@@ -155,8 +155,8 @@ void resetTopologyCache() {
                     if (!entry.isFiltered())
                         entries.add(new CacheContinuousQueryEvent<K, V>(cache, cctx, entry));
 
-                    if (log.isDebugEnabled())
-                        log.debug("Partition was lost [lastFiredEvt=" + lastFiredEvt +
+                    if (log.isInfoEnabled())
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryPartitionRecovery.java:158]: I suggest to revert this change. It's not related to the issue.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java [hunk 3]
```diff
@@ -376,34 +338,32 @@ private Batch initBatch(AffinityTopologyVersion topVer, boolean backup) {
...
+         * @param filteredFactory Factory which produces filtered entries.
+         * @return Map of collected entries.
          */
-        @Nullable synchronized TreeMap<Long, CacheContinuousQueryEntry> flushCurrentEntries(
-            @Nullable TreeMap<Long, CacheContinuousQueryEntry> res) {
-            if (entries == null)
+        synchronized Map<Long, CacheContinuousQueryEntry> flushCurrentEntries(
+            BiFunction<Long, Long, CacheContinuousQueryEntry> filteredFactory
+        ) {
+            Map<Long, CacheContinuousQueryEntry> res = new HashMap<>();
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java:364]: Empty map can be returned if `entries == null || filteredFactory == null`

Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java:364]: Can we estimate map size?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java (hunk 1 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java:101]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryPartitionRecovery.java (hunk 2 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryPartitionRecovery.java]: Reverted.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java (hunk 3 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java]: Agree, Fixed.

**Review by NSAmelchev — APPROVED** [review]
